Changing the mock server to use port reserver insted of hard ports

MeniZalzman opened the pull request:
Changing the Server mock to used port reserver instead of hard coding the port in the tests, This change also enable executing this tests in parallel as the port reserver making sure not to allocate reserved ports to other tests.

nuget/home#1050

@deepakaravindr @yishaigalatzer @emgarten @feiling

zhili1208 committed:
fixed https://github.com/NuGet/Home/issues/1052

Commit:
Changing the mock server to use port reserver insted of hard coding the ports in the tests

Commit:
Changing the mock server to use port reserver insted of hard coding the ports in the tests

dnfclas commented:
Hi **@MeniZalzman**, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla2.dotnetfoundation.org.
    </span>

TTYL, DNFBOT;

dnfclas commented:
**@MeniZalzman**, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, DNFBOT;

Commit:
Fixes https://github.com/NuGet/Home/issues/1029

Changed the dependency on NuGet.ISettings to NuGet.Configuration.ISettings.
Verified that all the existing NuGetPackCommand tests run successfully.
Also, explicitly verified that the new code is used for tests which involve
build during pack and uses the new NuGet.Configuration.ISettings

yishaigalatzer commented:
Does this change also include fixing all the consistently broken test you mentioned today?

@yishaigalatzer Yes all the tests are passing

yishaigalatzer commented on test/NuGet.CommandLine.Test/CommandRunner.cs:
```diff
@@ -12,7 +12,7 @@ public class CommandRunner
         // Item1 of the returned tuple is the exit code. Item2 is the standard output, and Item3 
         // is the error output.
-        public static Tuple<int, string, string> Run(string process, string workingDirectory, string arguments, bool waitForExit, int timeOutInMilliseconds = 60000,
+        public static Tuple<int, string, string> Run(string process, string workingDirectory, string arguments, bool waitForExit, int timeOutInMilliseconds = 600000000,
```
1. Please while you change code, make the lines shorter (no longer than 120 chars). It makes reviewing a PR much easier. The easy way to see this is in a github PR, you never want to scroll.
2. Why is the timeout changing? Unless there is a very good reason it shouldn't because a failed test will halt the build for a long time

MeniZalzman replied:
good catch, I change the timeout for debugging purposes, will revert this change

yishaigalatzer commented on test/NuGet.CommandLine.Test/MockServer.cs:
```diff
@@ -15,28 +15,29 @@ namespace NuGet.CommandLine.Test
+        public MockServer()
         {
-            _endPoint = endPoint;
+            this._portReserver = new PortReserver();
```
don't use `this.`

MeniZalzman replied:
Removed

yishaigalatzer commented on test/NuGet.CommandLine.Test/MockServer.cs:
```diff
@@ -286,7 +289,7 @@ void HandleRequest()
         {
             string nsAtom = "http://www.w3.org/2005/Atom";
-            var id = string.Format(CultureInfo.InvariantCulture, "{0}{1}", _endPoint, title);
+            var id = string.Format(CultureInfo.InvariantCulture, "{0}{1}", this.Uri, title);
```
this (and other places as well)

MeniZalzman replied:
Removed

yishaigalatzer commented on test/NuGet.CommandLine.Test/MockServer.cs:
```diff
@@ -340,6 +343,28 @@ public string ToOData(IPackage package)
+            }
+        }
+        
+        public void Dispose()
```
this pattern is unnecessary here, just use dispose.

MeniZalzman replied:
I prefer to use this dispose pattern

yishaigalatzer commented on test/NuGet.CommandLine.Test/NuGetDeleteCommandTest.cs:
```diff
@@ -85,33 +85,33 @@ public void DeleteCommand_DeleteFromHttpSource()
-                waitForExit: true);
-            server.Stop();
+                    "-Source", server.Uri + "nuget", "-NonInteractive" };
+                var r = CommandRunner.Run(
```
is there a race here?

MeniZalzman replied:
not that I see, where do you think we are expose to race condition?

yishaigalatzer commented on test/NuGet.CommandLine.Test/NuGetInstallCommandTest.cs:
```diff
@@ -456,23 +456,25 @@ public void InstallCommand_GetLastestPrereleaseVersion()
+                        workingDirectory,
+                        args,
+                        waitForExit: true);
+                    server.Stop();
```
new line above `server.stop` it looks like part of the Run this way

MeniZalzman replied:
Right because of that I missed it, no need to call server stop anymore it's part of the dispose method

yishaigalatzer commented on test/NuGet.CommandLine.Test/NuGetPushCommandTest.cs:
```diff
@@ -392,70 +396,72 @@ public void PushCommand_PushToServerBasicAuth()
-                    }
-                    else
+                        var h = r.Headers["Authorization"];
+                        var credential = System.Text.Encoding.Default.GetString(Convert.FromBase64String(h.Substring(6)));
```
drop unnecessary namespace and move to a using statement if possible

MeniZalzman replied:
done

yishaigalatzer commented on test/NuGet.CommandLine.Test/NuGetPushCommandTest.cs:
```diff
@@ -392,70 +396,72 @@ public void PushCommand_PushToServerBasicAuth()
+                    server.Start();
+
+                    // Act
+                    Environment.SetEnvironmentVariable("FORCE_NUGET_EXE_INTERACTIVE", "true");
```
this looks like a problem waiting to happen. There is a race between tests here. This should be protected by a mutex.

MeniZalzman replied:
@yishaigalatzer  Can we split this change to a different code review?

yishaigalatzer commented on test/NuGet.CommandLine.Test/NuGetPushCommandTest.cs:
```diff
@@ -483,70 +488,72 @@ public void PushCommand_PushToServerBasicAuthDisableBuffering()
+                    server.Start();
+
+                    // Act
+                    Environment.SetEnvironmentVariable("FORCE_NUGET_EXE_INTERACTIVE", "true");
```
same comment as above about potential race. Also requires proper cleanup

yishaigalatzer commented:
:watch:

Commit:
Address code review comments

Commit:
Address code review comments

Commit:
Set UserAgent.UserAgentString to use CommandLineConstants.UserAgent.

Related to https://github.com/NuGet/Home/issues/929

deepakaravindr commented:
Looks good to me

Commit:
Merge branch 'meniz/ChangingServerMockToUsePortReserver' of https://github.com/NuGet/NuGet.PackageManagement into meniz/ChangingServerMockToUsePortReserver
